File: rpkg/R/read_redatam.R
#' @title Read REDATAM dictionary
#' @description Read a DIC or DICX dictionary file directly into R.
#' @param dictionary A character string with the path to the dictionary file. It
#'  allows relative paths.
#' @return A list of tibbles, each one representing a table with the census
#'  entities (or levels) and their attributes (or variables).
#' @examples
#' # Read a dictionary file (Galapagos 2015)
#' zip <- system.file("extdata", "galapagos.zip", package = "redatam")
#' dout <- paste(tempdir(), "galapagos", sep = "/")
#' unzip(zip, exdir = dout)
#' read_redatam(paste(dout, "cg15.dic", sep = "/"))
#' @export
read_redatam <- function(dictionary) {
  dictionary <- normalizePath(dictionary)

  if (grepl("\\.dic$|\\.DIC$|\\.dicx$|\\.DICX$||\\.dicX", dictionary)) {
    res <- export_redatam_to_list_(dictionary)
  } else {
    stop("The file extension must be either .dic or .dicx")
  }

  message("Fixing encoding...")

  # 1. convert lists to data.tables
  res <- list_to_datatable_(res)
  res <- res[vapply(res, function(x) nrow(x) > 0, logical(1))]

  # 2. fix encoding (e.g., ""C\xf3digo Region" -> "Código Region")
  res <- fix_encoding_(res)

  message("Tidying names...")

  # 3. convert REGION, Region, Regi\u00f3n, etc. to region and so on
  res <- tidy_names_(res)

  message("Tidying variables...")

  # 4. remove leading and trailing whitespaces and multiple whitespaces
  res <- trim_and_clean_(res)

  # 6. harmonize column types
  res <- harmonize_types_(res)

  # 7. merge with their descriptions
  res <- merge_descriptions_(res)
  
  # 8. show as tibble
  res <- datatable_to_tibble_(res)

  # 9. put entities first
  res <- res[sort(names(res))]

  return(res)
}

File: pypkg/redatam/read_redatam.py
# redatam/read_redatam.py

from os.path import abspath
from pandas import DataFrame
from . import export_redatam_to_dict
from .utils import (
    dict_to_dataframe,
    fix_encoding,
    trim_and_clean,
    harmonize_types
)

def read_redatam(dictionary):
    dictionary = abspath(dictionary)

    if dictionary.lower().endswith(('.dic', '.dicx')):
        res = export_redatam_to_dict(dictionary)
    else:
        raise ValueError("The file extension must be either .dic or .dicx")

    print("Fixing encoding...")

    # 1. convert dictionaries to dataframes
    res = dict_to_dataframe(res)
    res = {k: v for k, v in res.items() if not v.empty}

    # 2. fix encoding (e.g., ""C\xf3digo Region" -> "Código Region")
    res = fix_encoding(res)

    print("Tidying variables...")

    # 3. remove leading and trailing whitespaces and multiple whitespaces
    res = trim_and_clean(res)

    # 4. harmonize column types
    res = harmonize_types(res)

    # 5. tidy entries

    # put dict names in lowercase
    res = {k.lower(): v for k, v in res.items()}

    # put colnames in lowercase
    res = {k: v.rename(columns=str.lower) for k, v in res.items()}

    # sort by dict name
    res = dict(sorted(res.items()))

    return res
